doc/man/man3/MainWindow.3
.TH "MainWindow" 3 "Jeudi 19 Mai 2016" "Version 1.0" "Urbanisme" \" -*- nroff -*-
.ad l
.nh
.SH NAME
MainWindow \- Fenêre principale de l'application\&. Chargée d'afficher la surface, les menus, les résultats et gérer le lancement des opérations\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <mainwindow\&.hpp>\fP
.PP
Est dérivée de QMainWindow\&.
.SS "Connecteurs publics"

.in +1c
.ti -1c
.RI "void \fBpopAbout\fP ()"
.br
.ti -1c
.RI "void \fBlaunchInit\fP ()"
.br
.ti -1c
.RI "void \fBlaunchResol\fP ()"
.br
.ti -1c
.RI "void \fBlaunchEval\fP ()"
.br
.ti -1c
.RI "void \fBresetField\fP ()"
.br
.in -1c
.SS "Fonctions membres publiques"

.in +1c
.ti -1c
.RI "\fBMainWindow\fP (QWidget *parent=0)"
.br
.ti -1c
.RI "\fBMainWindow\fP (unsigned nbCols, unsigned nbRows, unsigned serveDistance, unsigned roadsWidth, QWidget *parent=0)"
.br
.ti -1c
.RI "\fBMainWindow\fP (unsigned nbCols, unsigned nbRows, const \fBParameters\fP &params, QWidget *parent=0)"
.br
.ti -1c
.RI "void \fBaskSizes\fP ()"
.br
.ti -1c
.RI "void \fBaskParams\fP ()"
.br
.ti -1c
.RI "void \fBinitComponents\fP ()"
.br
.ti -1c
.RI "void \fBinitEvents\fP ()"
.br
.ti -1c
.RI "\fBField\fP & \fBget_initialField\fP ()"
.br
.in -1c
.SH "Description détaillée"
.PP 
Fenêre principale de l'application\&. Chargée d'afficher la surface, les menus, les résultats et gérer le lancement des opérations\&. 
.SH "Documentation des constructeurs et destructeur"
.PP 
.SS "MainWindow::MainWindow (QWidget * parent = \fC0\fP)\fC [explicit]\fP"
Constructeur par défaut 
.PP
\fBParamètres:\fP
.RS 4
\fIparent\fP 
.RE
.PP

.SS "MainWindow::MainWindow (unsigned nbCols, unsigned nbRows, unsigned serveDistance, unsigned roadsWidth, QWidget * parent = \fC0\fP)\fC [explicit]\fP"
Constructeur avec des valeurs initiales 
.PP
\fBParamètres:\fP
.RS 4
\fInbCols\fP Largeur de la surface 
.br
\fInbRows\fP Hauteur de la surface 
.br
\fIserveDistance\fP Distance de desserte de la résolution 
.br
\fIroadsWidth\fP Largeur des routes de la résolution 
.br
\fIparent\fP Widget parent 
.br
\fIparent\fP 
.RE
.PP

.SS "MainWindow::MainWindow (unsigned nbCols, unsigned nbRows, const \fBParameters\fP & params, QWidget * parent = \fC0\fP)\fC [explicit]\fP"
Constructeur avec des valeurs initiales 
.PP
\fBParamètres:\fP
.RS 4
\fInbCols\fP Largeur de la surface 
.br
\fInbRows\fP Hauteur de la surface 
.br
\fIparams\fP Paramètres de résolution 
.br
\fIparent\fP Widget parent 
.RE
.PP

.SH "Documentation des fonctions membres"
.PP 
.SS "void MainWindow::askParams ()"
Demande les paramètres du problème avec lesquels effectuer la résolution 
.SS "void MainWindow::askSizes ()"
Demande les tailles de la surface à traiter à l'utilisateur 
.SS "\fBField\fP& MainWindow::get_initialField ()\fC [inline]\fP"
Accesseur sur la surface initiale 
.PP
\fBRenvoie:\fP
.RS 4
initialField, une Surface, modifiable 
.RE
.PP

.SS "void MainWindow::initComponents ()"
Créer et instancie les composants graphiques 
.SS "void MainWindow::initEvents ()"
Créer les connections des signaux et slots de la fenêtre 
.SS "void MainWindow::launchEval ()\fC [slot]\fP"
Lance l'évaluation des deux objectifs TODO afficher les résultats dans la fenêtre plutôt que dans le terminal 
.SS "void MainWindow::launchInit ()\fC [slot]\fP"
Lance le placement des premières routes à partir des deux E/S 
.SS "void MainWindow::launchResol ()\fC [slot]\fP"
Lance la résolution du problème : Maximisation du nombre de routes exploitables, puis maximisation de l'accessibilité 
.SS "void MainWindow::popAbout ()\fC [slot]\fP"
Affiche la fenêtre de présentation 
.SS "void MainWindow::resetField ()\fC [slot]\fP"
Supprime toutes les routes autres que des E/S de la surface 

.SH "Auteur"
.PP 
Généré automatiquement par Doxygen pour Urbanisme à partir du code source\&.


src/Display/mainwindow.cpp
#include "mainwindow.hpp"

#include <QtGui/QDesktopServices>
#include <QtWidgets/QVBoxLayout>
#include <QtWidgets/QLabel>
#include <QtWidgets/QPushButton>
#include <QtWidgets/QAction>
#include <QtWidgets/QFileDialog>
#include <QtWidgets/QInputDialog>
#include <QtGui/QMouseEvent>

/// #########################
///      Constructeurs
/// #########################
//@{
MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent), initialField(0,0), parameters(0,0), fieldWidget(nullptr)
{
    //    askSizes();
    //    askParams();

    resolution= new Resolution(initialField, parameters);

    initComponents();
    initEvents();
}

MainWindow::MainWindow(unsigned nbCols, unsigned nbRows, unsigned serveDistance, unsigned roadsWidth, QWidget *parent)
    : QMainWindow(parent), initialField(nbCols, nbRows), parameters(serveDistance, roadsWidth)
{
    fieldWidget= new FieldWidget(&initialField, serveDistance);

    resolution= new Resolution(initialField, parameters);

    initComponents();
    initEvents();
}

MainWindow::MainWindow(unsigned nbCols, unsigned nbRows, const Parameters &params, QWidget *parent)
    : QMainWindow(parent), initialField(nbCols, nbRows), parameters(params)
{
    fieldWidget= new FieldWidget(&initialField, params.get_serve_distance());
    initialField.setUsables(params.get_serve_distance());

    resolution= new Resolution(*(fieldWidget->get_field()), parameters);

    initComponents();
    initEvents();
}

#define INITIAL_SCALE 32
void MainWindow::initComponents()
{
    aboutWidget= nullptr;

    resize(initialField.get_width() * INITIAL_SCALE , initialField.get_height() * INITIAL_SCALE +20);

    setCentralWidget(fieldWidget);
    fieldWidget->redraw();
    fieldWidget->show();

    // Menu bar
    aboutAction= menuBar()->addAction( tr("About") );
    initAction= menuBar()->addAction( tr("Initialise") );
    evalAction= menuBar()->addAction( tr("Evaluate") );
    QMenu* menu_resols= menuBar()->addMenu( tr("Resolutions") );
    usableAction= menu_resols->addAction( tr("Usable") );
    accessAction= menu_resols->addAction( tr("Access") );
    resetAction= menuBar()->addAction( tr("Reset") );
    flushAction= menuBar()->addAction( tr("Flush") );
    exportAction= menuBar()->addAction( tr("Export"));

    exportAction->setEnabled(false);
}

void MainWindow::initEvents()
{
    connect(aboutAction, &QAction::triggered, this, &MainWindow::popAbout);
    connect(initAction, &QAction::triggered, this, &MainWindow::launchInit);
    connect(evalAction, &QAction::triggered, this, &MainWindow::launchEval);
    connect(usableAction, &QAction::triggered, this, &MainWindow::launchLocalUsable);
    connect(accessAction, &QAction::triggered, this, &MainWindow::launchLocalAccess);
    connect(resetAction, &QAction::triggered, this, &MainWindow::resetField);
    connect(flushAction, &QAction::triggered, this, &MainWindow::emptyField);
    connect(exportAction, &QAction::triggered, this, &MainWindow::exportPareto);

    // Gestion du bouton export grisé ou non
    connect(usableAction, &QAction::triggered, exportAction, &QAction::setDisabled);
    connect(accessAction, &QAction::triggered, exportAction, &QAction::setDisabled);
    connect(initAction, &QAction::triggered, exportAction, &QAction::setEnabled);
    connect(resetAction, &QAction::triggered, exportAction, &QAction::setEnabled);
    connect(flushAction, &QAction::triggered, exportAction, &QAction::setEnabled);
}

//@}
/// #########################
///         Others
/// #########################
//@{

void MainWindow::updateWorkField()
{
    resolution->changeWorkField(fieldWidget->get_field());
}

//@}
/// #########################
///         Slots
/// #########################
//@{

void MainWindow::popAbout()
{
    if(aboutWidget == nullptr){
        aboutWidget = new QWidget();
        QVBoxLayout* VLay= new QVBoxLayout(aboutWidget);
        QLabel* signature= new QLabel("Florian DAVID");
        QLabel* about= new QLabel("Projet première année de master");
        QLabel* contact= new QLabel("Contact : <a href='mailto:f.david5@laposte.net?subject=Application Urbanisme'>f.david5@laposte.net</a>");
        QLabel* depot= new QLabel("Depot: <a href='https://github.com/flodavid/Urbanisme'>github.com/flodavid/Urbanisme</a>");
        QPushButton* valid= new QPushButton("OK", aboutWidget);
        VLay->addWidget(signature);
        VLay->addWidget(about);
        VLay->addWidget(contact);
        VLay->addWidget(depot);
        VLay->addWidget(valid);

        contact->setOpenExternalLinks(true);
        depot->setOpenExternalLinks(true);

        signature->setStyleSheet("QLabel {color : darkblue;  }");
        about->setStyleSheet("QLabel {color : darkblue;  }");
        contact->setStyleSheet("QLabel {color : darkblue;  }");
        depot->setStyleSheet("QLabel {color : darkblue;  }");
        connect(valid, SIGNAL(clicked()), aboutWidget, SLOT(close()) );
    }
    aboutWidget->show();
}

void MainWindow::launchInit()
{
    updateWorkField();

    Field& result_field= resolution->initResolution();

    fieldWidget->set_field(&result_field);
    fieldWidget->redraw();
    fieldWidget->show();
}

void MainWindow::launchLocalUsable()
{
    updateWorkField();

    QInputDialog widget_ask_nbMax/*= new QInputDialog*/(this);
    unsigned maxRoadsToAdd= (unsigned)widget_ask_nbMax.getInt(this, "Maximisation du nombre d'exploitables", "Nombre maximum de routes à ajouter ?\n0 pour aucune limite", 0);

    Field& result_field= resolution->localSearchUsableObjective(maxRoadsToAdd);

    fieldWidget->set_field(&result_field);

    fieldWidget->redraw();
    fieldWidget->show();
}

void MainWindow::launchLocalAccess()
{
    updateWorkField();

    QInputDialog widget_ask_nbMax/*= new QInputDialog*/(this);
    unsigned maxPathsToAdd= (unsigned)widget_ask_nbMax.getInt(this, "Maximisation accessibilité", "Nombre chemins à ajouter ?", 1);
//    delete widget_ask_nbMax;

    Field& result_field= resolution->localSearchAccessObjective(maxPathsToAdd);

    fieldWidget->set_field(&result_field);

    fieldWidget->redraw();
    fieldWidget->show();
}

void MainWindow::launchEval()
{
    resolution->evaluateBothObjectives();
}

void MainWindow::resetField()
{
    fieldWidget->set_field(&initialField);
    updateWorkField();

    std::cout << "Solution trouvée supprimée"<< std::endl;

    fieldWidget->redraw();
}

void MainWindow::emptyField()
{
    if (&initialField != fieldWidget->get_field()) {
//        delete fieldWidget->get_field();

        std::cout << "Solution trouvée supprimée"<< std::endl;
    }
    delete resolution;

    // Initialisation d'une nouvelle surface vide
    initialField= Field(initialField.get_width(), initialField.get_height());
    initialField.setUsables(parameters.get_serve_distance());

    fieldWidget->set_field(&initialField);
    resolution= new Resolution(initialField, parameters);

    fieldWidget->redraw();
}

void MainWindow::exportPareto()
{
    QFileDialog* file_browser= new QFileDialog(this);

    file_browser->setAcceptMode(QFileDialog::AcceptSave);
    file_browser->setNameFilter(tr("Save") +" (*.pareto.txt)");

    if(file_browser->exec() == QDialog::Accepted){
        std::string filename = file_browser->selectedFiles()[0].toStdString();
        std::cout <<"taille de "<< filename<< " : "<< filename.length() << std::endl;

        // Sauvegarde de la foret dans FireWidget qui effectue la procédure de Foret
        resolution->trySaveParetoToTxt(filename);

        exportAction->setEnabled(false);
    }
}

//@}
